crates/apollo-compiler/src/diagnostic.rs

```diff
@@ -0,0 +1,344 @@
+//! ```
+//!
+//! The [`Diagnostic`] type wraps errors that implement [`ToDiagnostic`] and provides
+//! the pretty-printing functionality. [`ToDiagnostic::to_diagnostic`] returns a diagnostic
+//! ready for formatting:
+//!
+//! ```rust
+//! # use apollo_compiler::{Schema, diagnostic::{ToDiagnostic, NodeLocation, CliReport}};
+//! # struct LintError {}
+//! # impl ToDiagnostic for LintError {
+//! #     fn location(&self) -> Option<NodeLocation> { None }
+//! #     fn report(&self, _report: &mut CliReport) {}
+//! # }
+//! fn print_errors(schema: &Schema, errors: &[LintError]) {
+//!     for error in errors {
+//!         // Debug-formatting uses colors.
+//!         eprintln!("{:?}", error.to_diagnostic(&schema.sources));
+//!     }
+//! }
+//! ```
+use crate::execution::GraphQLError;
+use crate::execution::GraphQLLocation;
+use crate::validation::FileId;
+use crate::SourceFile;
+use crate::SourceMap;
+use ariadne::ColorGenerator;
+use ariadne::ReportKind;
+use std::fmt;
+use std::io;
+use std::ops::Range;
+use std::sync::Arc;
+use std::sync::OnceLock;
+
+pub use crate::validation::NodeLocation;
+
+#[cfg(doc)]
+use crate::{ExecutableDocument, Schema};
+
+type MappedSpan = (FileId, Range<usize>);
+
+/// Translate a byte-offset location into a char-offset location for use with ariadne.
+fn map_span(sources: &SourceMap, location: NodeLocation) -> Option<MappedSpan> {
+    let source = sources.get(&location.file_id)?;
+    let mapped_source = source.mapped_source();
+    let start = mapped_source.map_index(location.offset());
+    let end = mapped_source.map_index(location.end_offset());
+    Some((location.file_id, start..end))
+}
+
+/// Provide a [`std::io::Write`] API for a [`std::fmt::Formatter`].
+struct WriteToFormatter<'a, 'b> {
+    f: &'a mut fmt::Formatter<'b>,
+}
+impl io::Write for WriteToFormatter<'_, '_> {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        let s = std::str::from_utf8(buf).map_err(|_| io::ErrorKind::Other)?;
+        self.f.write_str(s).map_err(|_| io::ErrorKind::Other)?;
+        Ok(buf.len())
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
+        Ok(())
+    }
+}
+
+/// A diagnostic report that can be printed to a CLI with pretty colors and labeled lines of
+/// GraphQL source code.
+///
+/// Custom errors can use this in their `Display` or `Debug` implementations to build a report and
+/// then write it out with [`fmt`].
+///
+/// [`fmt`]: CliReport::fmt
+pub struct CliReport {
+    sources: SourceMap,
+    colors: ColorGenerator,
+    report: ariadne::ReportBuilder<'static, MappedSpan>,
+}
+
+impl CliReport {
+    /// Returns a builder for creating diagnostic reports.
+    ///
+    /// Provide GraphQL source files and the main location for the diagnostic.
+    /// Source files can be obtained from [`Schema::sources`] or [`ExecutableDocument::sources`].
+    pub fn builder(sources: SourceMap, location: Option<NodeLocation>) -> Self {
```

Comment:
A report typically only exists transiently to call its `write` or `fmt` method, right? Could it own a `&'_ SourceMap` instead of `SourceMap`? (Or maybe I shouldn’t worry about micro-things like this…)

Similarly for `Diagnostic<T>`